internal/config/manager.go

```diff
@@ -0,0 +1,262 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package config
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/milvus-io/milvus/internal/log"
+	"go.uber.org/zap"
+)
+
+type Manager struct {
+	sync.RWMutex
+	sources      map[string]Source
+	keySourceMap sync.Map
+}
+
+func NewManager() *Manager {
+	return &Manager{
+		sources: make(map[string]Source),
+	}
+}
+
+func (m *Manager) GetConfig(key string) (string, error) {
+	realKey := formatKey(key)
+	sourceName, ok := m.keySourceMap.Load(realKey)
+	if !ok {
+		return "", fmt.Errorf("key not found: %s", key)
+	}
+	return m.configValueBySource(realKey, sourceName.(string))
+}
+
+// Configs returns all the key values
+func (m *Manager) Configs() map[string]string {
+	config := make(map[string]string)
+
+	m.keySourceMap.Range(func(key, value interface{}) bool {
+		sValue, err := m.configValueBySource(key.(string), value.(string))
+		if err != nil {
+			return true
+		}
+		config[key.(string)] = sValue
+		return true
+	})
+
+	return config
+}
+
+func (m *Manager) SetConfig(key, value string) {
+	realKey := formatKey(key)
+	m.sources[CustomSourceName].Set(realKey, value)
+	m.updateEvent(newEvent(CustomSourceName, CreateType, realKey, value))
+}
+
+func (m *Manager) DeleteConfig(key string) {
+	realKey := formatKey(key)
+	for source := range m.sources {
+		m.sources[source].Delete(realKey)
+		m.updateEvent(newEvent(realKey, DeleteType, realKey, ""))
+	}
+}
```

Comment:
SetConfig just change custom source, but  DeleteConfig remove from all source,  will it cause config lost that milvus role is relying on？

Reply:
Agree. Looks weird here. Actually all sources implement  `Set` and `Delete` with no-op.  Sources behave like immutable from the prospective of application. Shall we remove these two method from interface decl?

Reply:
you are right. the sources should be immutable by design, I do not want to add `Set` and `Delete` either 😅， but some of the tests rely on them. I try to find another way to make them work fine

Reply:
add an `overlay` map in `manager`, `set` operation will write into `overlay`, `delete` operation will set a `tombvalue` in `overlay`, when `GetConfig` will check whether the value is `tombvalue`, if true it will be judged as nil